java/tracing/src/main/scala/org/typelevel/otel4s/java/trace/SpanBuilderImpl.scala

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2022 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.java
+package trace
+
+import cats.effect.Resource
+import cats.effect.Sync
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import io.opentelemetry.api.trace.{Span => JSpan}
+import io.opentelemetry.api.trace.{SpanBuilder => JSpanBuilder}
+import io.opentelemetry.api.trace.{SpanKind => JSpanKind}
+import io.opentelemetry.api.trace.{Tracer => JTracer}
+import io.opentelemetry.context.{Context => JContext}
+import org.typelevel.otel4s.Attribute
+import org.typelevel.otel4s.trace.Span
+import org.typelevel.otel4s.trace.SpanBuilder
+import org.typelevel.otel4s.trace.SpanContext
+import org.typelevel.otel4s.trace.SpanFinalizer
+import org.typelevel.otel4s.trace.SpanKind
+
+import scala.concurrent.duration.FiniteDuration
+
+private[trace] final case class SpanBuilderImpl[F[_]: Sync](
+    jTracer: JTracer,
+    name: String,
+    scope: TraceScope[F],
+    parent: SpanBuilderImpl.Parent = SpanBuilderImpl.Parent.Propagate,
+    finalizationStrategy: SpanFinalizer.Strategy =
+      SpanFinalizer.Strategy.reportAbnormal,
+    kind: Option[SpanKind] = None,
+    links: Seq[(SpanContext, Seq[Attribute[_]])] = Nil,
+    attributes: Seq[Attribute[_]] = Nil,
+    startTimestamp: Option[FiniteDuration] = None
+) extends SpanBuilder[F] {
+
+  import SpanBuilderImpl._
+
+  def withSpanKind(spanKind: SpanKind): SpanBuilder[F] =
+    copy(kind = Some(spanKind))
+
+  def withAttribute[A](attribute: Attribute[A]): SpanBuilder[F] =
+    copy(attributes = attributes :+ attribute)
+
+  def withAttributes(attributes: Attribute[_]*): SpanBuilder[F] =
```

Comment:
I think I expected this to replace, not append, attributes.  I'd have called this `addAttributes`.  That has ramifications for some others (`withAttribute`, `withLink`).

`withFinalizationStrategy` doesn't append.

Reply:
`addX` fits better indeed. I updated the names.